File: c++实用代码/network flow/ditch/src/zhjshar1_ditch/ditch/ditch.cpp
/*
ID:zhjshar1
PROG:ditch
LANG:C++
*/
#include <fstream>
using namespace std;

ifstream  in( "ditch.in" );
ofstream out( "ditch.out" );

const int N = 3011, OO = 2000000010;
int c[N][N]={0}, f[N][N] = {0}, nv, sq, tq;
inline int Ziabs( int x ){
       return ( x >= 0 ? x : -x );
}
void Input( void );          // c, nv, sq, tq
void Ff( void );
void Output( void );

int main(){
    Input();
    Ff();
    Output();
    return 0;
}

void Input( void ){
     int ne, i, j, k, tc;
     in >>ne >>nv;
     sq=1; tq=nv;
     for( k=0; k<ne; ++k ){
          in >>i >>j >>tc;
          c[i][j] += tc;
     }
     return;
}

void Ff( void ){
     int pre[N], d, q[N], h, t, i, j, k;
     bool inq[N];
     pre[tq] = 1;
     while( pre[tq] != 0 ){
            memset( pre, 0, sizeof(pre) );
            memset( inq, 0, sizeof(inq) );
            h = t = 0;
            q[h] = sq;
            inq[sq] = true;
            while( (h<=t) && (pre[tq]==0) ){
                   i = q[h];
                   for( j=nv; j>0; --j )
                        if( !inq[j] ){
                            if( (c[i][j]) && (f[i][j]<c[i][j]) ){
                                q[++t] = j;
                                inq[j] = true;
                                pre[j] = i;
                            }
                            else if( (c[j][i]) && (f[j][i]>0) ){
                                 q[++t] = j;
                                 inq[j] = true;
                                 pre[j] = -i;
                            }
                        }
                   h++;
            }
                        
            if( pre[tq] != 0 ){
                d = OO;
                j = tq; i = Ziabs( pre[j] );
                while( i ){
                       if( pre[j] > 0 ) k = c[i][j] - f[i][j];
                       else             k = f[j][i];
                       if( k < d ) d = k;
                       j = i; i = Ziabs( pre[j] );
                }
                
                j = tq; i = Ziabs( pre[j] );
                while( i ){
                       if( pre[j] > 0 ) f[i][j] += d;
                       else             f[j][i] -= d;
                       j = i; i = Ziabs( pre[j] );
                }
            }
     }
     return;
}

void Output( void ){
     int maxflow = 0;
     for( int i=nv; i>0; --i )
          maxflow += f[sq][i];
     out <<maxflow <<endl;
     return;
}

File: c++实用代码/network flow/ditch/src2/dfs网络流_ydl/ditch/ditch.pas
{
ID:yudingl2
PROG:ditch
LANG:PASCAL
}

var
 n,m,i,j,x,y,z,ans,min:longint;
 a:array[1..2000,1..2000] of longint;
 pre:array[1..2000] of longint;
 f:array[1..2000] of boolean;
 flag:boolean;

procedure dfs(x:longint);
var i:longint;
begin
 if x=n then
  begin
   flag:=false;
   exit;
  end;
 f[x]:=false;
 for i:=1 to n do
  begin
   if f[i] and (a[x,i]<>0) and flag then
    begin
     pre[x]:=i;
     dfs(i);
     if not flag then exit;
    end;
   if not flag then exit;
  end;
end;

procedure network_flow;
var i,j,k,L:longint;
begin
 while true do
  begin
   flag:=true;
   fillchar(f,sizeof(f),true);
   fillchar(pre,sizeof(pre),0);
   dfs(1);
   if flag then break;
   min:=maxlongint;
   k:=1;L:=pre[k];
   while L<>0 do
    begin
     if a[k,L]<min then min:=a[k,L];
     k:=L;L:=pre[k];
    end;
   k:=1;L:=pre[k];
   while L<>0 do
    begin
     a[k,L]:=a[k,L]-min;
     a[L,k]:=a[L,k]+min;
     k:=L;L:=pre[k];
    end;
   ans:=ans+min;
  end;
end;

begin
 assign(input,'ditch.in');reset(input);
 assign(output,'ditch.out');rewrite(output);
 randomize;
 readln(m,n);
 for i:=1 to m do
  begin
   readln(x,y,z);
   inc(a[x,y],z);
//inc(a[y,x],z);
  end;
for i:=1 to n do a[i][i]:=0;
{for i:=1 to n do
  for j:=1 to n do
  if (a[i][j]>0)and(a[j][i]>0)then writeln('wa');}
 network_flow;
 writeln(ans);
 close(input);close(output);
end.
